### PR TITLE
Scala 3 support [superseded by #2687]

### LINKED ISSUE #2177: Scala 3 support
Stefan made an initial stab at it in https://github.com/slick/slick/pull/2123, which was never merged

a volunteer is needed to rebase and resubmit it, updating it from an old Dotty version to a current Scala 3 version

### PR DESCRIPTION
**Pull request by alexFrankfurt**
Continuation of #2123.
Closes #2177.

3.0.0-M1
- [x] root / compile
- [x] testkit / test

3.0.0-M2
- [x] root / compile
- [x] testkit / test (1 Skipped)

(3.2.2, 2.13.10) ~~(1c217b7)~~ ([2c071f1](https://github.com/slick/slick/pull/2187/commits/6c14ffc9e5f614e54912c9173b1c5e6c3caa1c11c))
- [x] root / compile 
- [ ] Enable back project  `reactive-streams-tests`
- [ ] Enable back tests:  for db2 {} DB2Test, MySQLTest
- [ ] Enable back '-Wunused:imports'
- [ ] Enable back tests:
  - [ ] com.typesafe.slick.testkit.tests.ActionTest#testSessionPinning
  - [ ] com.typesafe.slick.testkit.tests.ThreeWayJoinTest#testManyToManyJoin [gist](https://gist.github.com/alexFrankfurt/4b9fd5bd647d7c48979993d9542eb089)
  - [ ] com.typesafe.slick.testkit.tests.ActionTest#testCreateIfNotExistsDropIfExists[gist](https://gist.github.com/alexFrankfurt/b98a33397fcaec07141891310bb6ca3c)
  - [ ] com.typesafe.slick.testkit.tests.ModelBuilderTest#test [gist](https://gist.github.com/alexFrankfurt/14515f071561815180cdc87df8274e4e)
  - [ ] com.typesafe.slick.testkit.tests.ThreeWayJoinTest#testManyToManyJoinTwice [gist](com.typesafe.slick.testkit.tests.ThreeWayJoinTest#testManyToManyJoinTwice)
  - [ ] com.typesafe.slick.testkit.tests.ActionTest#testSimpleActionAsFuture [gist](https://gist.github.com/alexFrankfurt/8074be973c817680f7051deb35871733)
  - [ ] com.typesafe.slick.testkit.tests.ActionTest#testOptionSequence [gist](https://gist.github.com/alexFrankfurt/95a7b4f2c1109d81b8c659bff93dbf33)
  - [ ] com.typesafe.slick.testkit.tests.ActionTest#testFlatten  com.typesafe.slick.testkit.tests.ActionTest#testZipWith com.typesafe.slick.testkit.tests.ActionTest#testCollect com.typesafe.slick.testkit.tests.ActionTest#testTruncate [gist](https://gist.github.com/alexFrankfurt/0c1ce248b5132a9ab56fc1c3e865dda8)
  - [ ] com.typesafe.slick.testkit.tests.ActionTest#testStreaming [gist](https://gist.github.com/alexFrankfurt/bcb405092678f6c5c5f3ed0f40533acf)
  - [ ] com.typesafe.slick.testkit.tests.ColumnDefaultTest#test
- [ ] testkit / test
`[error] Failed: Total 1349, Failed 22, Errors 0, Passed 1327, Ignored 1`

Differences with main:
~~1. Samples removed from GitHub Actions tests~~
2. sbt-scoverage plugin removed (for this PR, to run tests for all versions in GitHub Actions)
3. project `reactive-streams-tests` temporarily disabled

### CODE REVIEW

**Comment by SethTisue** [comment]
There were transient failures on CI. I restarted the build. (But our Travis-CI is pretty backed up so we might not get results until tomorrow.)

**Comment by SethTisue** [comment]
the failures this time are real — looks like you dropped 2.11 support (which is fine, since it seems highly likely that #2184, which does that too, will be merged) but missed a place where the build or CI config still refers to it

**Comment by SethTisue** [comment]
@alexFrankfurt #2184 is merged now, so you could merge or rebase this now and see if CI goes green?

**Comment by izmailoff** [comment]
Just some feedback. It compiles with `scalaVersion` 3.0.0-M1 and passes all tests on my machine. 3.0.0-M2 doesn't compile simply because files like `./slick/src/main/scala-3.0.0-M2/slick/lifted/ShapedValue.scala` do not exist.

Although I'm able to compile, I can't do `publishLocal` because doc generation fails. It gives lots of 'bad option' warnings and can't unpickle some classes. I haven't looked deeper, but perhaps this needs to be also updated to use Scala 3 (reflection problems?).

**Comment by SethTisue** [comment]
>  I can't do publishLocal because doc generation fails

This problem will very likely go away once the target version is 3.0.0-RC2.

**Comment by nafg** [comment]
What's the next step for this PR?

**Comment by alexFrankfurt** [comment]
@nafg making all the necessary code changes to compile with latest release 3.0.0-RC2, and after, see what is needed for Travic CI to build successfully.

@izmailoff version specific code goes in directory with version in name, therefore paths need to be updated to reflect scalaVersion like her: https://github.com/slick/slick/pull/2187/commits/3f44d7dc7948b05e82233b731835fa7500a4eef9#diff-188343fd1224095c8114c44f6e3471019c2779951af584381d902fd71960cfc5

next looking what can be done to usages of case class 's companion object 's .tupled method

**Comment by nafg** [comment]
I think we need to make `.mapTo` smarter.

Even in Scala 2, it's an issue once I define my own companion object for a case class. If the case class parameter list is more complicated than just (Int, Int) it's annoying to make the companion object extend the appropriate function type. What I usually do is add this in the companion:

```scala
  def tupled = (apply _).tupled
```

But fixing `mapTo` would be a much better idea.

Of course, since it's a macro fixing it means fixing it twice, once for the scala 2 macro and once for the scala 3 macro, I guess.

Speaking of which, how dichotomous is the codebase going to be now? Is it going to be complicated to make any refactoring as long as it's cross-compiling between scala 2 and scala 3?

**Comment by SethTisue** [comment]
https://github.com/slick/slick/pull/2202 is merged, so at next rebase you can drop sbt-dotty from plugins.sbt

**Comment by nafg** [comment]
Is this targeting 3.4.0 or 4.0.0?

Also, if there were a separate PR that removed scala 2.12 and fixed deprecations, would it be a lot of work to rebase this onto it?

**Comment by nafg** [comment]
Dropping 2.12 actually has major ramifications. It means it can't be used within SBT, including slick-codegen.

**Comment by osleonard** [comment]
is anyone working on completing this, otherwise I could start taking a look at whats left

**Comment by nafg** [comment]
@osleonard I don't think so, although @xuwei-k has been doing some tasks, so that would be great.

Note that there have been some steps towards scala 3 compatibility on master (e.g. #2240, #2243), and other changes, so one helpful task would be rebasing this branch onto `main`.

Some other useful tasks, besides getting this branch to work fully on 3.0.0:

 - Fix existing warnings like inner class "overriding" (see https://twitter.com/not_xuwei_k/status/1406144015200882689)
 - Add scala option `-Xsource:3` (should work on 2.12.14 and 2.13.6) and follow the resulting warnings and/or errors
 - Figure out what breaking changes might be possible to mitigate

Whatever could be done in separate PRs is better IMO.

**Comment by alexFrankfurt** [comment]
@nafg , @osleonard 
Currently paused work on this PR after commit https://github.com/slick/slick/pull/2187/commits/1c217b75ee5fd4e39cf54ec5fc75311ff5512dc5.
Tests run successfully locally for 2.13.5 and 3.0.0.

**Comment by nafg** [comment]
@alexFrankfurt that's great thanks! What remains to be done and what are your plans?

**Comment by alexFrankfurt** [comment]
@nafg 
There are several failing tests in testkit, need to be fixed. They differ from failures in Github Actions.
Same for 2.13.5 and 3.0.0: `[error] Failed: Total 1288, Failed 9, Errors 0, Passed 1278, Skipped 1`
And conflicts with several merged PRs that need to be analyzed for further decisions.
https://gist.github.com/alexFrankfurt/3c2ce2077e56cbdb53011ede7b4a5771

I am gonna be off for several weeks.

**Comment by osleonard** [comment]
> @nafg
> There are several failing tests in testkit, need to be fixed. They differ from failures in Github Actions.
> Same for 2.13.5 and 3.0.0: `[error] Failed: Total 1288, Failed 9, Errors 0, Passed 1278, Skipped 1`
> And conflicts with several merged PRs that need to be analyzed for further decisions. (Scoverage, samples)
> https://gist.github.com/alexFrankfurt/3c2ce2077e56cbdb53011ede7b4a5771
> 
> I am gonna be off for several weeks.

@alexFrankfurt will start to take a look at whats left from Monday morning.

**Comment by romainreuillon** [comment]
Dear devs,

I am looking forward to the slick release for scala 3. Would you have any update about it?

best,
Romain

**Comment by nafg** [comment]
Me to. Do you want to help move it forward?

**Comment by osleonard** [comment]
Hey guys, had my hands full the past weeks / month, having some free time from weekend so I will get to it

**Comment by robstoll** [comment]
@nafg I can try to move things forward. Does it still make sense to:
(a) rebase this PR on main  
(b) move any other changes that don't break anything to their own PR that can be merged sooner. 

or do we need to wait for another change first?
Also, are you still willing to do pair programming? I have not yet contributed to slick and it would surely help me.

**Comment by nafg** [comment]
> @nafg I can try to move things forward. Does it still make sense to: (a) rebase this PR on main

Yes

> (b) move any other changes that don't break anything to their own PR that can be merged sooner.

Yes

> 
> or do we need to wait for another change first? Also, are you still willing to do pair programming?

And yes! ;)

There are some other things needed to actually merge Scala 3 support, but they shouldn't hold up getting it working.

**Comment by robstoll** [comment]
@nafg sorry, I actually wanted to inform you right after your response that I won't have time to work on this until next week. Anyway... I will start to work on this next week :slightly_smiling_face: Is there a way to contact you privately (say slack, gitter or...) in case I have questions and to find a time slot where we could do a pair programming session.

**Comment by nafg** [comment]
> @nafg sorry, I actually wanted to inform you right after your response that I won't have time to work on this until next week. Anyway... I will start to work on this next week :slightly_smiling_face:

Great!

> Is there a way to contact you privately (say slack, gitter or...) in case I have questions and to find a time slot where we could do a pair programming session.

I can't give you Slack contact without knowing what Slack workspaces we are both on. You can DM me on gitter (search my github handle), but its notifications aren't great. If you're on Discord you can DM me as `nafg#1263`.

Once you DM me I could give you other options.

**Comment by robstoll** [comment]
@nafg I have sent you a friend request on discord, looks like you have disabled DM for non-friends.

**Comment by akozumpl** [comment]
The set removes `tsql` support, are we fine with that for now?

**Comment by akozumpl** [comment]
No answer, I take that that we're not sure.

I think there's a lot of interest in making this happen but the set is long and has (at least) this question mark around it. So what I wanted to propose is getting through the PR in multiple pieces, merging in the stuff we know we're sure about. Would that be desirable or at least acceptable?

**Comment by robstoll** [comment]
@akozumpl I am currently cherry-picking non scala 3 changes from the dottyquery branch so that those coukd already be merged (see #2404). You could take that task over if you like. If you want, then I am going to document what commits I skipped etc. How does that sound to you?

**Comment by akozumpl** [comment]
Hi @robstoll , that sounds good! what's the next step then?

**Comment by nafg** [comment]
I pushed the nafg/dottyquery branch which is a rebase on `main`. I'm not sure if I broke anything or not. It needs some history-rewriting work still.

**Comment by akozumpl** [comment]
> the nafg/dottyquery branch

`sbt test` passes here, actually much faster:
```
[info] Passed: Total 1325, Failed 0, Errors 0, Passed 1325, Ignored 1
[success] Total time: 229 s (03:49), completed Apr 3, 2022, 10:22:46 AM
```

vs current `main`:
```
[info] Passed: Total 1364, Failed 0, Errors 0, Passed 1364
[success] Total time: 340 s (05:40), completed Apr 3, 2022, 10:12:25 AM
```

I guess that's good.

File: slick-codegen/src/main/scala/slick/codegen/OutputHelpers.scala [hunk 1]
```diff
@@ -112,8 +112,11 @@ trait ${container}Root {
 package ${pkg}
 // AUTO-GENERATED Slick data model
 /** Stand-alone Slick data model for immediate use */
-object ${container} extends ${container} {
-  val profile = $profile
+
+object ${container} extends ${container}_objectBase with ${container}
+
```
Comment (nafg) [inline slick-codegen/src/main/scala/slick/codegen/OutputHelpers.scala:117]: @szeiger can you please share the reason for this change?

**Comment by nafg** [comment]
Discussion https://github.com/slick/slick/discussions/2515

**Comment by takapi327** [comment]
I tried the Scala3-supported Slick with MySQL and got a stack overflow.

The [expr](https://github.com/slick/slick/pull/2187/files#diff-776df962449d573db74a1079763acd22d5652df82b6458d32007bc8937c02742R316) method is created with the same naming and arguments as the existing one, and it is changed to call the expr method of the process that has been performed so far.
In this case, if the expr method is overwritten in the inheritance destination and the newly created expr method is called, the expr method called in the process will call the one overwritten by itself, resulting in an infinite loop.

If there is a condition match within an expr method that has been overridden in the inherited destination that does not match, change it so that the newly created non-expr method of the direct inheritance source is called.

Since other classes that inherit from QueryBuilder seemed to have similar fixes, it is assumed that only MySQL omitted to support this.

<details>
<summary> Details </summary>

## Version used
- Scala 3.2.0
- Slick 3.4.0-SNAPSHOT

## Executable code

```scala
case class Person(
  id:   Option[Long],
  name: String,
  age:  Option[Int],
)

class PersonTable(tag: Tag) extends Table[Person](tag, "person"):
  def id   = column[Long]("id", O.PrimaryKey, O.AutoInc)
  def name = column[String]("name")
  def age  = column[Option[Int]]("age")

  def * = (id.?, name, age).mapTo[Person]

object Example:

  private val dataSource = ???

  private val personTable = TableQuery[PersonTable]
  private val db = Database.forDataSource(dataSource, None)

  def main(args: Array[String]): Unit =
    val query = for {
      p <- personTable.filter(_.name === "takapi327")
    } yield p.name
    val result = db.run(query.result.headOption)
    result.map(println(_))

    Await.result(result, Duration.Inf)
    db.close
```

## Error Description
```shell
[error] Exception in thread "main" java.lang.StackOverflowError
[error]         at slick.jdbc.MySQLProfile$MySQLQueryBuilder.expr(MySQLProfile.scala:216)
[error]         at slick.jdbc.JdbcStatementBuilderComponent$QueryBuilder.expr$$anonfun$1(JdbcStatementBuilderComponent.scala:317)
[error]         at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error]         at slick.jdbc.JdbcStatementBuilderComponent$QueryBuilder.withSkipParens(JdbcStatementBuilderComponent.scala:125)
[error]         at slick.jdbc.JdbcStatementBuilderComponent$QueryBuilder.expr(JdbcStatementBuilderComponent.scala:317)
[error]         at slick.jdbc.MySQLProfile$MySQLQueryBuilder.expr(MySQLProfile.scala:228)
...
[error] Nonzero exit code returned from runner: 1
```
</details>

**Comment by nafg** [comment]
Replaced by #2687

**Comment by nafg** [comment]
@takapi327 can you try the newer branch? If it's still an issue it would be helpful if you can open a PR adding it as a test case

**Comment by takapi327** [comment]
@nafg 
Thanks!
I tried with a new branch and had the same problem.
alexFrankfurt:Please check back when you have a chance, as we have created a new pull-link to avoid the inclusion of dottyquery-specific differences.

https://github.com/slick/slick/pull/2689

**Comment by nafg** [comment]
Welcome back @alexFrankfurt ;) Meanwhile I'd been working on a different approach (rebasing and splitting commits) but I'm not sure it's the best one. Any chance you'd be open to having a call sometime this week?

**Comment by alexFrankfurt** [comment]
@nafg yes, you can contact me on discord alexFrankfurt#4120

**Comment by alexFrankfurt** [comment]
Implemented in #2687

**Pull request closed without merging**